### PR TITLE
refactor auth

### PR DESCRIPTION
**Pull request by lunny**
- [x] Rename `models/login` to `models/auth`
- [x] Move auth related routers into a standalone sub package
- [x] Move `security` setting files into a sub `security` folder
- [x] Move `security_*.tmpl` into `security/*.tmpl`

### CODE REVIEW

**Review by KN4CK3R — APPROVED** [review]
Unit test fail seems related but I don't know how. Looks like `db.DefaultContext` is `nil`? Otherwise lgtm.

File: models/db/convert.go [hunk 1]
```diff
@@ -39,3 +42,16 @@ func ConvertUtf8ToUtf8mb4() error {
...
 }
+
+// Cell2Int64 converts a xorm.Cell type to int64,
+// and handles possible irregular cases.
+func Cell2Int64(val xorm.Cell) int64 {
+	switch (*val).(type) {
+	case []uint8:
+		log.Trace("Cell2Int64 ([]uint8): %v", *val)
+
+		v, _ := strconv.ParseInt(string((*val).([]uint8)), 10, 64)
```
Comment (lafriks) [inline models/db/convert.go:53]: should error really be ignored?

Comment (lunny) [reply models/db/convert.go:53]: Just move from `models`, haven't take a look at it. :)

Comment (lunny) [reply models/db/convert.go:53]: It could be ignored, even if the error catch, it has to return zero.

File: routers/web/auth/2fa.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+
+		if ctx.Session.Get("linkAccount") != nil {
+			gothUser := ctx.Session.Get("linkAccountGothUser")
+			if gothUser == nil {
+				ctx.ServerError("UserSignIn", errors.New("not in LinkAccount session"))
+				return
+			}
```
Comment (wxiaoguang) [inline routers/web/auth/2fa.go]: These lines are newly added? Should we use `LinkAccountFromStore` ?

File: routers/web/auth/oauth.go [hunk 3]
```diff
@@ -750,3 +757,253 @@ func handleAuthorizeError(ctx *context.Context, authErr AuthorizeError, redirect
...
+func oAuth2UserLoginCallback(authSource *auth.Source, request *http.Request, response http.ResponseWriter) (*user_model.User, goth.User, error) {
+	gothUser, err := authSource.Cfg.(*oauth2.Source).Callback(request, response)
+	if err != nil {
+		if err.Error() == "securecookie: the value is too long" || strings.Contains(err.Error(), "Data too long") {
+			log.Error("OAuth2 Provider %s returned too long a token. Current max: %d. Either increase the [OAuth2] MAX_TOKEN_LENGTH or reduce the information returned from the OAuth2 provider", authSource.Name, setting.OAuth2.MaxTokenLength)
+			err = fmt.Errorf("OAuth2 Provider %s returned too long a token. Current max: %d. Either increase the [OAuth2] MAX_TOKEN_LENGTH or reduce the information returned from the OAuth2 provider", authSource.Name, setting.OAuth2.MaxTokenLength)
+		}
+		return nil, goth.User{}, err
+	}
+
```
Comment (wxiaoguang) [inline routers/web/auth/oauth.go:1070]: A lot of code are lost. By purpose?


```

	if oauth2Source.RequiredClaimName != "" {
		claimInterface, has := gothUser.RawData[oauth2Source.RequiredClaimName]
		if !has {
			return nil, goth.User{}, user_model.ErrUserProhibitLogin{Name: gothUser.UserID}
		}

		if oauth2Source.RequiredClaimValue != "" {
			groups := claimValueToStringSlice(claimInterface)
			found := false
			for _, group := range groups {
				if group == oauth2Source.RequiredClaimValue {
					found = true
					break
				}
			}
			if !found {
				return nil, goth.User{}, user_model.ErrUserProhibitLogin{Name: gothUser.UserID}
			}
		}
	}
```

File: routers/web/auth/oauth.go [hunk 4]
```diff
@@ -750,3 +757,253 @@ func handleAuthorizeError(ctx *context.Context, authErr AuthorizeError, redirect
...
+		}
+		if err := ctx.Session.Release(); err != nil {
+			log.Error("Error storing session: %v", err)
+		}
+
+		// Clear whatever CSRF has right now, force to generate a new one
+		middleware.DeleteCSRFCookie(ctx.Resp)
+
+		// Register last login
+		u.SetLastLogin()
```
Comment (wxiaoguang) [inline routers/web/auth/oauth.go:989]: Lost code:
```
		// Update GroupClaims
		changed := setUserGroupClaims(source, u, &gothUser)
		cols := []string{"last_login_unix"}
		if changed {
			cols = append(cols, "is_admin", "is_restricted")
		}
```

File: routers/web/auth/oauth.go [hunk 5]
```diff
@@ -750,3 +757,253 @@ func handleAuthorizeError(ctx *context.Context, authErr AuthorizeError, redirect
...
+		if redirectTo := ctx.GetCookie("redirect_to"); len(redirectTo) > 0 {
+			middleware.DeleteRedirectToCookie(ctx.Resp)
+			ctx.RedirectToFirst(redirectTo)
+			return
+		}
+
+		ctx.Redirect(setting.AppSubURL + "/")
+		return
+	}
+
```
Comment (wxiaoguang) [inline routers/web/auth/oauth.go:1022]: Lost code:
```

	changed := setUserGroupClaims(source, u, &gothUser)
	if changed {
		if err := user_model.UpdateUserCols(db.DefaultContext, u, "is_admin", "is_restricted"); err != nil {
			ctx.ServerError("UpdateUserCols", err)
			return
		}
	}
```

File: routers/web/auth/oauth.go (hunk 3 above)
Comment (lunny) [reply routers/web/auth/oauth.go:1070]: Accident.

**Comment by lunny** [comment]
> Some comments are broken by replacing `login.` to empty string.
> 
> image

done.

File: routers/web/auth/2fa.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	if ok && twofa.LastUsedPasscode != form.Passcode {
+		remember := ctx.Session.Get("twofaRemember").(bool)
+		u, err := user_model.GetUserByID(id)
+		if err != nil {
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+
+		if ctx.Session.Get("linkAccount") != nil {
+			err = externalaccount.LinkAccountFromStore(ctx.Session, ctx.User)
```
Comment (wxiaoguang) [inline routers/web/auth/2fa.go]: Old code is `externalaccount.LinkAccountFromStore(ctx.Session, u)`.

Since they should works at the same way, but I prefer to keep the old logic during refactoring.

File: routers/web/auth/2fa.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+		}
+
+		remember := ctx.Session.Get("twofaRemember").(bool)
+		u, err := user_model.GetUserByID(id)
+		if err != nil {
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+
+		handleSignInFull(ctx, u, remember, false)
```
Comment (wxiaoguang) [inline routers/web/auth/2fa.go:156]: Lost code:

```
		if ctx.Written() {
			return
		}
```

File: routers/web/auth/auth.go [hunk 8]
```diff
@@ -0,0 +1,795 @@
...
+	if err := ctx.Session.Release(); err != nil {
+		log.Error("Unable to store session: %v", err)
+	}
+
+	// Language setting of the user overwrites the one previously set
+	// If the user does not have a locale set, we save the current one.
+	if len(u.Language) == 0 {
+		u.Language = ctx.Locale.Language()
+		if err := user_model.UpdateUserCols(db.DefaultContext, u, "language"); err != nil {
+			log.Error(fmt.Sprintf("Error updating user language [user: %d, locale: %s]", u.ID, u.Language))
```
Comment (wxiaoguang) [inline routers/web/auth/auth.go]: The old code is 
```
			ctx.ServerError("UpdateUserCols Language", fmt.Errorf("Error updating user language [user: %d, locale: %s]", u.ID, u.Language))
```

I think it should not be changed during refactoring.

File: routers/web/auth/linkaccount.go [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+		LoginSource: authSource.ID,
+		LoginName:   gothUser.UserID,
+	}
+
+	if !createAndHandleCreatedUser(ctx, tplLinkAccount, form, u, &gothUser, false) {
+		// error already handled
+		return
+	}
+
+	ctx.Redirect(setting.AppSubURL + "/user/login")
```
Comment (wxiaoguang) [inline routers/web/auth/linkaccount.go]: Old code is `handleSignIn(ctx, u, false)`, why is this changed?

File: routers/web/auth/password.go [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+		return
+	}
+
+	if len(form.Password) < setting.MinPasswordLength {
+		ctx.Data["Err_Password"] = true
+		ctx.RenderWithErr(ctx.Tr("auth.password_too_short", setting.MinPasswordLength), tplMustChangePassword, &form)
+		return
+	}
+
+	var err error
```
Comment (wxiaoguang) [inline routers/web/auth/password.go]: A lot of code lost:

```
	if !password.IsComplexEnough(form.Password) {
		ctx.Data["Err_Password"] = true
		ctx.RenderWithErr(password.BuildComplexityError(ctx), tplMustChangePassword, &form)
		return
	}
	pwned, err := password.IsPwned(ctx, form.Password)
	if pwned {
		ctx.Data["Err_Password"] = true
		errMsg := ctx.Tr("auth.password_pwned")
		if err != nil {
			log.Error(err.Error())
			errMsg = ctx.Tr("auth.password_pwned_err")
		}
		ctx.RenderWithErr(errMsg, tplMustChangePassword, &form)
		return
	}
```

File: routers/web/auth/password.go [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+		if err = auth.UpdateTwoFactor(twofa); err != nil {
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+
+		handleSignInFull(ctx, u, remember, false)
+		ctx.Flash.Info(ctx.Tr("auth.twofa_scratch_used"))
```
Comment (wxiaoguang) [inline routers/web/auth/password.go:260]: Lost code:
```
		if ctx.Written() {
			return
		}
```

File: routers/web/auth/password.go [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+		ctx.Redirect(setting.AppSubURL + "/user/settings/security")
+		return
+	}
+
+	handleSignInFull(ctx, u, remember, true)
```
Comment (wxiaoguang) [inline routers/web/auth/password.go]: Old code is `	handleSignIn(ctx, u, remember)`

File: routers/web/auth/u2f.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+			}
+			redirect := handleSignInFull(ctx, user, remember, false)
+			if redirect == "" {
+				redirect = setting.AppSubURL + "/"
+			}
+			ctx.PlainTextBytes(200, []byte(redirect))
+			return
+		}
+	}
+	ctx.Error(http.StatusUnauthorized)
```
Comment (wxiaoguang) [inline routers/web/auth/u2f.go:135]: Old code is:

```
			if ctx.Session.Get("linkAccount") != nil {
				if err := externalaccount.LinkAccountFromStore(ctx.Session, user); err != nil {
					ctx.ServerError("UserSignIn", err)
				}
			}
			redirect := handleSignInFull(ctx, user, remember, false)
			if ctx.Written() {
				return
			}
			if redirect == "" {
				redirect = setting.AppSubURL + "/"
			}
			ctx.PlainText(http.StatusOK, redirect)
			return
		}
	}
	ctx.Error(http.StatusUnauthorized)
```

**Review by wxiaoguang — CHANGES_REQUESTED** [review]

File: routers/web/auth/2fa.go (hunk 6 above)
Comment (lunny) [reply routers/web/auth/2fa.go]: Good catch!

**Comment by lunny** [comment]
@wxiaoguang I think all done.

**Review by wxiaoguang — APPROVED** [review]
